akka-actor/src/main/scala/akka/io/DirectByteBufferPool.scala

```diff
@@ -54,11 +54,42 @@ private[akka] class DirectByteBufferPool(defaultBufferSize: Int, maxPoolEntries:
     }
   }
 
-  private final def offerBufferToPool(buf: ByteBuffer): Unit =
-    pool.synchronized {
-      if (buffersInPool < maxPoolEntries) {
-        pool(buffersInPool) = buf
-        buffersInPool += 1
-      } // else let the buffer be gc'd
+  private final def offerBufferToPool(buf: ByteBuffer): Unit = {
+    val clean =
+      pool.synchronized {
+        if (buffersInPool < maxPoolEntries) {
+          pool(buffersInPool) = buf
+          buffersInPool += 1
+          false
+        } else {
+          // try to clean it outside the lock, or let the buffer be gc'd
+          true
+        }
+      }
+    if (clean)
+      tryCleanDirectByteBuffer(buf)
+  }
+
+  /**
+   * DirectByteBuffers are garbage collected by using a phantom reference and a
+   * reference queue. Every once a while, the JVM checks the reference queue and
+   * cleans the DirectByteBuffers. However, as this doesn't happen
+   * immediately after discarding all references to a DirectByteBuffer, it's
+   * easy to OutOfMemoryError yourself using DirectByteBuffers. This function
+   * explicitly calls the Cleaner method of a DirectByteBuffer.
+   *
+   * Utilizes reflection to avoid dependency to `sun.misc.Cleaner`.
+   */
+  private final def tryCleanDirectByteBuffer(toBeDestroyed: ByteBuffer): Unit = try {
+    if (toBeDestroyed.isDirect) {
+      val cleanerMethod = toBeDestroyed.getClass().getMethod("cleaner")
```

Comment:
1.On a jvm implementation which don't have `UNSAFE`/`Cleaner` thing,this implement cause much `throw` and `catch`. 
2. We could first determine whether the runtime have the  `UNSAFE`/`Cleaner` ,if it have one,we then get the offset of the `Cleaner` in the `DirectByteBuffer`，after that,we could use the `UNSAFE` to get the `Cleaner` and there is no need for the reflection.